core/templates/domain/objects/RatioObjectFactory.ts

```diff
@@ -0,0 +1,110 @@
+// Copyright 2020 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Factory for creating instances of Ratio
+ * domain objects.
+ */
+
+import { downgradeInjectable } from '@angular/upgrade/static';
+import { Injectable } from '@angular/core';
+
+import { ObjectsDomainConstants } from
+  'domain/objects/objects-domain.constants';
+import { RatioInputAnswer } from
+  'interactions/answer-defs';
+
+export class Ratio {
+  numbers: number[];
+  constructor(numbersList: number[]) {
+    this.numbers = numbersList;
+  }
+
+  toString(): string {
+    return this.numbers.join(':');
+  }
+
+  getNumberOfTerms(): number {
+    return this.numbers.length;
+  }
+  convertToSimplestForm(): number[] {
+    var gcd = (x: number, y: number) => {
+      return y === 0 ? x : gcd(y, x % y);
+    };
+    var gcdResult = this.numbers.reduce(gcd);
+    return this.numbers.map(currentValue => {
+      if (currentValue === 0) {
+        return currentValue;
+      } else {
+        return currentValue / gcdResult;
+      }
+    });
+  }
+}
+
+@Injectable({
+  providedIn: 'root'
+})
+export class RatioObjectFactory {
+  fromRawInputString(rawInput: string): number[] {
+    if (rawInput.length === 0) {
+      throw new Error(
+        ObjectsDomainConstants.RATIO_PARSING_ERRORS.EMPTY_STRING);
+    }
+    var INVALID_CHARS_REGEX = /[^\d^:^\.^\/^\s]/g;
+    if (INVALID_CHARS_REGEX.test(rawInput)) {
+      throw new Error(
+        ObjectsDomainConstants.RATIO_PARSING_ERRORS.INVALID_CHARS);
+    }
+    var INVALID_COLON_REGEX = /(:{2})/g;
+    if (INVALID_COLON_REGEX.test(rawInput)) {
+      throw new Error(
+        ObjectsDomainConstants.RATIO_PARSING_ERRORS.INVALID_COLONS);
+    }
+    var INVALID_RATIO_REGEX = /\d+[\.\/]{1,}\d*/g;
+    if (INVALID_RATIO_REGEX.test(rawInput)) {
+      throw new Error(
+        ObjectsDomainConstants.RATIO_PARSING_ERRORS.NON_INTEGER_ELEMENTS);
```

Comment:
we could do that as well but in this case, we wanted to specific about the fraction or decimal values.

Reply:
I suggest checking for all non integer cases here, unless that’s covered elsewhere. It will make the overall checking more robust. For example, what error will be shown for the input string: “1$:4”?

Reply:
$ is an invalid sign so invalid character error would be thrown here.